cirq/contrib/two_qubit_gates/gate_compilation.py

```diff
@@ -0,0 +1,328 @@
+                                             ignore_equivalent_vectors=True)
+        nearest_ind = infidelities.argmin()
+
+        success = infidelities[nearest_ind] < self.max_expected_infidelity
+
+        # shape (n,2,2,2)
+        inner_gates = np.array(self.single_qubit_gates[nearest_ind])
+
+        if inner_gates.size == 0:  # Only need base gate
+            outer_locals, actual = _outer_locals_for_unitary(
+                unitary, self.base_gate)
+            return outer_locals, actual, success
+
+        # reshape to operators on 2 qubits, (n,4,4)
+        inner_gates = vector_kron(inner_gates[..., 0, :, :],
+                                  inner_gates[..., 1, :, :])
+
+        assert inner_gates.ndim == 3
+        inner_product = reduce(lambda a, b: self.base_gate @ b @ a, inner_gates,
+                               self.base_gate)
+        outer_locals, actual = _outer_locals_for_unitary(unitary, inner_product)
+
+        out = [outer_locals[0]]
+        out.extend(self.single_qubit_gates[nearest_ind])
+        out.append(outer_locals[1])
+
+        return out, actual, success
+
+
+def _outer_locals_for_unitary(
+        target: np.ndarray, base: np.ndarray
+) -> Tuple[Tuple[_SingleQubitGatePair, _SingleQubitGatePair], np.ndarray]:
+    """Local unitaries mapping between locally equivalent 2-local unitaries.
+
+    Finds the left and right 1-local unitaries k_L, k_R such that
+
+    U_target = k_L @ U_base @ k_R
+
+    Args:
+        target: The unitary to which we want to map.
+        base: The base unitary which maps to target.
+
+    Returns:
+        (k_R, k_L) : The right and left 1-local unitaries in the equation above,
+            expressed as 2-tuples of (2x2) single qubit unitaries.
+        actual: The outcome of k_L @ base @ k_R
+    """
+    target_decomp = kak_decomposition(target)
+    base_decomp = kak_decomposition(base)
+
+    # From the KAK decomposition, we have
+    # kLt At kRt = kL kLb Ab KRb kR
+    # If At=Ab, we can solve for the right and left local unitaries as
+    # kLt = kL kLb --> kL = kLt kLb^\dagger
+    # kRt = kRb kR --> kR = kRb\dagger kRt
+
+    kLt0, kLt1 = target_decomp.single_qubit_operations_after
+    kLb0, kLb1 = base_decomp.single_qubit_operations_after
+    k_L = kLt0 @ kLb0.conj().T, kLt1 @ kLb1.conj().T
+
+    kRt0, kRt1 = target_decomp.single_qubit_operations_before
+    kRb0, kRb1 = base_decomp.single_qubit_operations_before
+    k_R = kRb0.conj().T @ kRt0, kRb1.conj().T @ kRt1
+
+    actual = np.kron(*k_L) @ base
+    actual = actual @ np.kron(*k_R)
+    actual *= np.conj(target_decomp.global_phase)
+
+    return (k_R, k_L), actual
+
+
+def _tabulate_KAK_vectors(
+        tabulation: Dict[int, Tuple[_SingleQubitGatePair, ...]],
+        base_gate: np.ndarray,
+        max_dist: float,
+        KAK_mesh: np.ndarray,
+        *local_unitary_pairs: _SingleQubitGatePair,
+) -> Tuple[List[np.ndarray], List[Tuple[_SingleQubitGatePair, ...]]]:
```

Comment:
I notice you're using names to refer to the output values in the return doc. Change this return type into a `NamedTuple`.

Reply:
Done